layer_factory/assistant_layer/mem_bound_warning.cpp

```diff
@@ -0,0 +1,210 @@
+                                                   VkDeviceSize memoryOffset, VkResult result) {
+    if (result == VK_SUCCESS) SetBindMemoryState<VkBuffer>(buffer, memory);
+
+    return VK_SUCCESS;
+}
+
+VkResult MemBoundWarning::PostCallBindBufferMemory2(VkDevice device, uint32_t bindInfoCount,
+                                                    const VkBindBufferMemoryInfo *pBindInfos, VkResult result) {
+    if (result == VK_SUCCESS) {
+        for (uint32_t i = 0; i < bindInfoCount; i++) {
+            SetBindMemoryState<VkBuffer>(pBindInfos[i].buffer, pBindInfos[i].memory);
+        }
+    }
+
+    return VK_SUCCESS;
+}
+
+VkResult MemBoundWarning::PostCallBindBufferMemory2KHR(VkDevice device, uint32_t bindInfoCount,
+                                                       const VkBindBufferMemoryInfo *pBindInfos, VkResult result) {
+    if (result == VK_SUCCESS) {
+        for (uint32_t i = 0; i < bindInfoCount; i++) {
+            SetBindMemoryState<VkBuffer>(pBindInfos[i].buffer, pBindInfos[i].memory);
+        }
+    }
+
+    return VK_SUCCESS;
+}
+
+// 3 buffer memory state tracker entry points
+VkResult MemBoundWarning::PostCallBindImageMemory(VkDevice device, VkImage image, VkDeviceMemory memory, VkDeviceSize memoryOffset,
+                                                  VkResult result) {
+    if (result == VK_SUCCESS) SetBindMemoryState<VkImage>(image, memory);
+
+    return VK_SUCCESS;
+}
+
+VkResult MemBoundWarning::PostCallBindImageMemory2(VkDevice device, uint32_t bindInfoCount,
+                                                   const VkBindImageMemoryInfoKHR *pBindInfos, VkResult result) {
+    if (result == VK_SUCCESS) {
+        for (uint32_t i = 0; i < bindInfoCount; i++) {
+            SetBindMemoryState<VkImage>(pBindInfos[i].image, pBindInfos[i].memory);
+        }
+    }
+
+    return VK_SUCCESS;
+}
+
+VkResult MemBoundWarning::PostCallBindImageMemory2KHR(VkDevice device, uint32_t bindInfoCount,
+                                                      const VkBindImageMemoryInfo *pBindInfos, VkResult result) {
+    if (result == VK_SUCCESS) {
+        for (uint32_t i = 0; i < bindInfoCount; i++) {
+            SetBindMemoryState<VkImage>(pBindInfos[i].image, pBindInfos[i].memory);
+        }
+    }
+
+    return VK_SUCCESS;
+}
+
+// intercept for the sparse memory to track memory state
+VkResult MemBoundWarning::PostCallQueueBindSparse(VkQueue queue, uint32_t bindInfoCount, const VkBindSparseInfo *pBindInfo,
+                                                  VkFence fence, VkResult result) {
+    for (uint32_t bindIdx = 0; bindIdx < bindInfoCount; bindIdx++) {
+        const VkBindSparseInfo &bindInfo = pBindInfo[bindIdx];
+
+        for (uint32_t j = 0; j < bindInfo.bufferBindCount; j++) {
+            for (uint32_t k = 0; k < bindInfo.pBufferBinds[j].bindCount; k++) {
+                auto sparse_binding = bindInfo.pBufferBinds[j].pBinds[k];
```

Comment:
It looks like this is making a copy. Did you intend to use a reference instead? Same for the two cases immediately below.

Reply:
I agree. Fixed in upcoming commit